utils/metrics/metrics.go

```diff
@@ -0,0 +1,187 @@
+package metrics
+
+import (
+	"github.com/buchgr/bazel-remote/config"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"strings"
+)
+
+// TODO Add test cases for this file
+
+type Method int
+type Status int
+type Kind int
+type Protocol int
+
+const (
+	METHOD_GET Method = iota
+	METHOD_HEAD
+	METHOD_PUT
+	METHOD_OTHER
+)
+
+const (
+	OK Status = iota
+	NOT_FOUND
+	OTHER_STATUS
+)
+
+const (
+	AC Kind = iota
+	CAS
+)
+
+const (
+	HTTP Protocol = iota
+	GRPC
+)
+
+func (e Method) String() string {
+	// Actually HTTP names, but can be conceptually mapped also to GRPC protocol.
+	if e == METHOD_GET {
+		return "GET"
+	}
+	if e == METHOD_HEAD {
+		return "HEAD"
+	}
+	if e == METHOD_PUT {
+		return "PUT"
+	}
+	return "other"
+}
+
+func (e Status) String() string {
+	// Names that works for both HTTP and GRPC, instead of HTTP or GRPC specific codes.
+	if e == OK {
+		return "OK"
+	}
+	if e == NOT_FOUND {
+		return "NotFound"
+	}
+	return "other"
+}
+
+func (e Kind) String() string {
+	if e == AC {
+		return "AC"
+	}
+	if e == CAS {
+		return "CAS"
+	}
+	return "other"
+}
+
+type Metrics interface {
+	// TODO Document interface
+	IncomingRequestCompleted(kind Kind, method Method, status Status, headers map[string][]string, protocol Protocol)
+}
+
+type metrics struct {
+	categoryValues               map[string]map[string]struct{}
+	counterIncomingCompletedReqs *prometheus.CounterVec
+}
+
+func NewMetrics(config *config.Metrics) Metrics {
+
+	labels := []string{"method", "status", "kind"}
+	categoryValues := make(map[string]map[string]struct{})
+
+	if config != nil && config.Categories != nil {
+		for categoryName, whiteListedValues := range config.Categories {
```

Comment:
Let's try to use the term "allowedValues", it's more inclusive.

Reply:
yes, good point!